kms.go

```diff
@@ -0,0 +1,186 @@
+package hdfs
+
+import (
+	"bytes"
+	"crypto/aes"
+	"io/ioutil"
+
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	hdfs "github.com/colinmarc/hdfs/v2/internal/protocol/hadoop_hdfs"
+	spnego "github.com/jcmturner/gokrb5/v8/spnego"
+)
+
+const (
+	kmsSchemeHTTP  = "kms://http@"
+	kmsSchemeHTTPS = "kms://https@"
+)
+
+func (c *Client) kmsAuth(url string) error {
+	if c.options.KerberosServicePrincipleName == "" {
+		url += ("&user.name=" + c.options.User)
+	}
+
+	req, err := http.NewRequest("OPTIONS", url, nil)
+	if err != nil {
+		return err
+	}
+
+	var resp *http.Response
+	if c.options.KerberosClient != nil {
+		kHttp := spnego.NewClient(c.options.KerberosClient, c.http, "")
```

Comment:
Let's put the kms stuff into a `kmsClient` and then we can reuse stuff like this.

Reply:
You mean to create `KmsClient` class, with all stuff for KMS interactions ?